Pcap++/header/KniDevice.h

```diff
@@ -0,0 +1,709 @@
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_network_if)(uint16_t port_id, uint8_t if_up);
+			/**
+			 * Pointer to function of configuring mac address.
+			 * If callback is not provided and port_id of KNI device is not UINT16_MAX
+			 * then DPDK will use predefined callback witch sets MAC address of
+			 * DPDK port port_id via rte_eth_dev_default_mac_addr_set.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_mac_address)(uint16_t port_id, uint8_t mac_addr[]);
+			/**
+			 * Pointer to function of configuring promiscuous mode.
+			 * If callback is not provided and port_id of KNI device is not UINT16_MAX
+			 * then DPDK will use predefined callback witch sets promiscuous mode of
+			 * DPDK port port_id via rte_eth_promiscuous_enable/rte_eth_promiscuous_disable.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_promiscusity)(uint16_t port_id, uint8_t to_on);
+		};
+
+		/**
+		 * @brief Old callbacks for KNI device events.
+		 * This structure MUST be used ONLY when KniDevice#callbackVersion returns
+		 * KniCallbackVersion#CALLBACKS_OLD.
+		 * Or if You are sure that DPDK version used is lower than 17.11.
+		 * If some callback is not provided (NULL) the request will always succeeds.
+		 */
+		struct KniOldIoctlCallbacks
+		{
+			/**
+			 * Pointer to function of changing MTU.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*change_mtu)(uint8_t port_id, unsigned int new_mtu);
+			/**
+			 * Pointer to function of configuring network interface.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_network_if)(uint8_t port_id, uint8_t if_up);
+		};
+
+		/**
+		 * @brief KNI device initialization data.
+		 * Used to create new KNI device.
+		 * Usage of callbacks member or oldCallbacks member is defined by
+		 * result of KniDevice#callbackVersion
+		 */
+		struct KniDeviceConfiguration
+		{
+			enum
+			{
+				// Must correspond to minimum of
+				// RTE_KNI_NAMESIZE [DPDK] = 32 and
+				// IFNAMSIZ [LINUX] = 16 (on most systems)
+				KNI_NAME_SIZE = 16
+			};
+			/**
+			 * Name used to display device in system.
+			 * Must not interfere with already existing network interfaces.
+			 */
+			char name[KNI_NAME_SIZE];
+			union
+			{
+				KniIoctlCallbacks* callbacks;
+				KniOldIoctlCallbacks* oldCallbacks;
+			};
+			/**
+			 * Pointer to MAC (ETHERNET) address of new KNI device.
+			 * If omitted (NULL) some valid address will be automatically generated.
+			 * If provided will be cached by new KNI device info structure.
+			 */
+			MacAddress* mac;
+			/**
+			 * Used in same way as DPDK port id.
+			 * If some new callbacks are omitted then have separate meaning
+			 * (see KniIoctlCallbacks#config_mac_address and KniIoctlCallbacks#config_promiscusity).
+			 * Can be set to UINT16_MAX.
+			 */
+			uint16_t portId;
```

Comment:
It is not that simple. DPDK will act differently on ocasion if no event callbacks are set for this KNI device. It will try to apply actions like changing the MAC to DPDK device which have the portid same as this KNI dvice. So i think the user must control it by himself. 
> what happens if the user defines 2 KNI devices with the same id?

They are created sucessfully if they have different names. Like stated in description of this variable the value `UINT16_MAX` is safe to use regarding the DPDK behavior described above.
The `getDeviceByPort` function in case of two KNI devices with same port will yield some of them (which one is not defined)

Reply:
my suggestion would be that if you let the user decide on portId, at least put some checks that the user can't define 2 devices with the same ID, it'll reduce confusion

Reply:
I think this will be great. For the obvious reasons stated above I will exclude `UINT16_MAX`  from that checks.

Reply:
awesome! 👍

Reply:
\+